### PR TITLE
Small optimizations

### PR DESCRIPTION
**Pull request by snapshotpl**

### CODE REVIEW

File: src/CallbackStream.php [hunk 1]
```diff
@@ -172,10 +172,10 @@ public function getMetadata($key = null)
             return $metadata;
         }
 
-        if (! array_key_exists($key, $metadata)) {
-            return null;
+        if (isset($metadata[$key])) {
```
Comment (Ocramius) [inline src/CallbackStream.php]: `return $metadata[$key] ?? null;`

File: src/RelativeStream.php [hunk 2]
```diff
@@ -107,7 +107,7 @@ public function isSeekable()
      */
     public function seek($offset, $whence = SEEK_SET)
     {
-        if ($whence == SEEK_SET) {
+        if ($whence === SEEK_SET) {
```
Comment (Ocramius) [inline src/RelativeStream.php]: Can you use the FQN for this constant?

File: src/ServerRequestFactory.php [hunk 3]
```diff
@@ -60,7 +60,7 @@ public static function fromGlobals(
         $files   = static::normalizeFiles($files ?: $_FILES);
         $headers = static::marshalHeaders($server);
 
-        if (null === $cookies && array_key_exists('cookie', $headers)) {
+        if (null === $cookies && isset($headers['cookie'])) {
```
Comment (Ocramius) [inline src/ServerRequestFactory.php]: You need to introduce tests around `null` values to avoid regressions here

File: src/ServerRequestFactory.php [hunk 4]
```diff
@@ -206,7 +206,7 @@ public static function marshalHeaders(array $server)
 
                 // We will not overwrite existing variables with the
                 // prefixed versions, though
-                if (array_key_exists($key, $server)) {
+                if (isset($server[$key])) {
                     continue;
```
Comment (Ocramius) [inline src/ServerRequestFactory.php]: You need to introduce tests around `null` values to avoid regressions here

File: src/Stream.php [hunk 5]
```diff
@@ -285,11 +285,11 @@ public function getMetadata($key = null)
         }
 
         $metadata = stream_get_meta_data($this->resource);
-        if (! array_key_exists($key, $metadata)) {
-            return null;
+        if (isset($metadata[$key])) {
```
Comment (Ocramius) [inline src/Stream.php]: Use `??` here too

File: src/Uri.php [hunk 6]
```diff
@@ -641,7 +641,7 @@ private function filterQuery($query)
     private function splitQueryValue($value)
     {
         $data = explode('=', $value, 2);
-        if (1 === count($data)) {
+        if (! isset($data[1])) {
```
Comment (Ocramius) [inline src/Uri.php]: This check is not equivalent to the previous: you need to verify that the key `0` exists too

File: src/CallbackStream.php (hunk 1 above)
Comment (snapshotpl) [reply src/CallbackStream.php]: Good idea but after drop php5 ;(

File: src/RelativeStream.php (hunk 2 above)
Comment (snapshotpl) [reply src/RelativeStream.php]: I'm not sure what are you meant

Comment (Ocramius) [reply src/RelativeStream.php]: `\SEEK_SET`

File: src/Uri.php (hunk 6 above)
Comment (samsonasik) [reply src/Uri.php]: @Ocramius `explode()` should always has key 0 ? https://3v4l.org/viNZT

File: src/ServerRequestFactory.php (hunk 3 above)
Comment (snapshotpl) [reply src/ServerRequestFactory.php]: I think it's imposible to write this test becase `$headers` comes from private function which alware return strings pair array.

File: src/Uri.php (hunk 6 above)
Comment (snapshotpl) [reply src/Uri.php]: @Ocramius agree with @samsonasik

File: src/CallbackStream.php (hunk 1 above)
Comment (Ocramius) [reply src/CallbackStream.php]: Wha?! This is for an improvement: just drop it already

File: src/ServerRequestFactory.php (hunk 3 above)
Comment (Ocramius) [reply src/ServerRequestFactory.php]: No need then: nullability was just an assumption on my end.

File: src/Uri.php (hunk 6 above)
Comment (Ocramius) [reply src/Uri.php]: Yup, misread

File: src/CallbackStream.php (hunk 1 above)
Comment (snapshotpl) [reply src/CallbackStream.php]: But is it not better to merge it as it to current php5 supported line and then, drop php5 and apply you have suggested?

**Comment by snapshotpl** [comment]
@Ocramius all remarks applied in code or I miss something?

**Comment by webimpress** [comment]
@snapshotpl tests are failing. Why?

File: src/RelativeStream.php [hunk 7]
```diff
@@ -105,9 +103,9 @@ public function isSeekable()
     /**
      * {@inheritdoc}
      */
-    public function seek($offset, $whence = SEEK_SET)
+    public function seek($offset, $whence = \SEEK_SET)
     {
-        if ($whence == SEEK_SET) {
+        if ($whence === \SEEK_SET) {
```
Comment (webimpress) [inline src/RelativeStream.php]: I would go with import the constant as we do in other places for internal functions:
```php
use const SEEK_SET;
```

File: src/ServerRequestFactory.php (hunk 4 above)
Comment (weierophinney) [reply src/ServerRequestFactory.php]: Agreed, particularly as the values may resolve to a `false` (e.g., `0`, `''`). The reason `array_key_exists()` was used was to allow _any_ value, regardless of whether or not it evaluated as truthy.

File: src/Stream.php [hunk 8]
```diff
@@ -193,11 +195,11 @@ public function isWritable()
...
-            strstr($mode, 'x')
-            || strstr($mode, 'w')
-            || strstr($mode, 'c')
-            || strstr($mode, 'a')
-            || strstr($mode, '+')
+            strpos($mode, 'x') !== false
+            || strpos($mode, 'w') !== false
+            || strpos($mode, 'c') !== false
+            || strpos($mode, 'a') !== false
+            || strpos($mode, '+') !== false
```
Comment (weierophinney) [inline src/Stream.php]: How is this change "better" or an "optimization", exactly? I find it less readable.

File: src/Stream.php [hunk 9]
```diff
@@ -234,7 +236,7 @@ public function isReadable()
         $meta = stream_get_meta_data($this->resource);
         $mode = $meta['mode'];
 
-        return (strstr($mode, 'r') || strstr($mode, '+'));
+        return (strpos($mode, 'r') !== false || strpos($mode, '+') !== false);
```
Comment (weierophinney) [inline src/Stream.php]: Same comment as above regarding `strstr` vs `strpos`. I recognize that `strpos` is marginally faster, but most folks agree it's a micro-optimization at best. If using it sacrifices readability, I'm not sure the change is useful.

File: src/Stream.php (hunk 5 above)
Comment (weierophinney) [reply src/Stream.php]: If you target PHP 7.1+, you'll need to update the `composer.json`, rebase against develop, and change the branch this is against.

File: src/CallbackStream.php (hunk 1 above)
Comment (weierophinney) [reply src/CallbackStream.php]: If we can optimize for PHP 5.6 today, we can make it even better for the next _major_ release later. I'm fine with this for now.

**Comment by weierophinney** [comment]
Closing; many changes were requested, but none provided by the author.

If you or anyone else reading want to see these changes, please submit a new PR incorporating them, along with the requested feedback.

Thanks!
